masonry/src/passes/update.rs

```diff
@@ -402,6 +403,19 @@ pub(crate) fn run_update_focus_pass(root: &mut RenderRoot) {
     }
 
     let prev_focused = root.global_state.focused_widget;
+    let was_ime_active = root.global_state.is_ime_active;
+
+    let synthesize_ime_disabled =
+        was_ime_active && prev_focused != root.global_state.next_focused_widget;
+    if synthesize_ime_disabled {
+        // IME was active, but the next focused widget is going to receive the
+        // Ime::Disabled event. Synthesize an Ime::Disabled event here and send it to
+        // the widget about to be unfocused.
+        run_on_text_event_pass(root, &TextEvent::Ime(winit::event::Ime::Disabled));
+    }
```

Comment:
Should we emit the EndIme signal here?

Reply:
If this is hit, the `EndIme` signal is emitted in the refocus steps below (line 507). I'm happy with moving it up. There may be some more opportunities to clean this code up a bit; there appear at least to be some invariants that the current code misses, for example, `was_ime_active` appears to imply `prev_focused.is_some()`, and `is_ime_active` implies `next_focused.is_some()`.

Reply:
I've simplified IME in the focus pass a bit, behavior should be the same.